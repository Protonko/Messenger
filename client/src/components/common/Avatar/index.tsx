import {useMemo, FC} from 'react'
import classNames from 'classnames'
import {Sizes} from 'models/common/sizes'
import {ColorGeneratorByName} from 'utils/ColorGeneratorByName'
import {isDarkColor} from 'utils/isDarkColor'
import {Loader} from 'components/common/Loader'

export interface IPropsAvatar {
  src: string
  name?: string
  size?: Sizes
  customStyles?: string
}

export const Avatar: FC<IPropsAvatar> = ({
  src,
  name,
  size = Sizes.MEDIUM,
  customStyles,
}) => {
  const backgroundColor = useMemo(
    () => new ColorGeneratorByName(name ?? '').generate(),
    [src, name],
  )
  const avatarClassname = classNames(
    'avatar',
    {'avatar--sm': size === Sizes.SMALL},
    {'avatar--lg': size === Sizes.LARGE},
    {[customStyles ?? '']: customStyles},
  )

  if (!src && name) {
    const classNamesText = classNames(
      'avatar__text',
      {'avatar__text--light': isDarkColor(backgroundColor)},
      {'avatar__text--dark': !isDarkColor(backgroundColor)},
    )

    return (
      <div className={avatarClassname} style={{backgroundColor}}>
        <span className={classNamesText}>{name.charAt(0)}</span>
      </div>
    )
  }

  return (
    <div className={avatarClassname}>
      <img className="avatar__image" src={src} alt="" />
      <Loader />
    </div>
  )
}
