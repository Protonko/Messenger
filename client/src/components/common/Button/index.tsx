import {FC, HTMLProps, ReactNode} from 'react'

interface IButtonProps extends HTMLProps<HTMLButtonElement> {
  text?: string
  icon?: ReactNode
  leftAddons?: ReactNode
  rightAddons?: ReactNode
  type?: 'button' | 'submit' | 'reset'
}

export const Button: FC<IButtonProps> = ({
  text,
  icon,
  leftAddons,
  rightAddons,
  ...buttonProps
}) => {
  const renderButtonContent = () => {
    return (
      <>
        {text && (
          <span className="button__text">
            {text}
          </span>
        )}
        {icon && (
          <span className="button__icon">
            {icon}
          </span>
        )}
      </>
    )
  }

  return (
    <button className="button" {...buttonProps}>
      {leftAddons}
      {renderButtonContent()}
      {rightAddons}
    </button>
  )
}
