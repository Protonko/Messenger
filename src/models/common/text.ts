export enum TextTypes {
  h1 = 'h1',
  h2 = 'h2',
  h3 = 'h3',
  h4 = 'h4',
  h5 = 'h5',
  h6 = 'h6',
  p = 'p',
  mixed = 'mixed',
}

export enum TextWeight {
  LIGHT = 'LIGHT',
  REGULAR = 'REGULAR',
  MEDIUM = 'MEDIUM',
  BOLD = 'BOLD',
}

export enum TextSize {
  EXTRA_SMALL= 'XS',
  SMALL = 'SM',
  MEDIUM = 'MD',
  BIG = 'LG',
  LARGE = 'XL',
  EXTRA_LARGE = 'XXL',
}
