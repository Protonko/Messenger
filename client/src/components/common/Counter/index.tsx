// types
import {Status} from 'models/common/status'

import {FC} from 'react'
import classNames from 'classnames'

export interface IPropsCounter {
  count: number | '99+',
  status?: Status,
}

export const Counter: FC<IPropsCounter> = ({
  count,
  status = Status.ACTIVE,
}) => {
  const classNameStatus = classNames(
    'counter',
    {'counter--muted': status === Status.MUTED}
  )

  return (
    <div className={classNameStatus}>
      <span className="counter__text">
        {count}
      </span>
    </div>
  )
}
