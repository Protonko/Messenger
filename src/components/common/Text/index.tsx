// types
import {TextTypes, TextSize, TextWeight} from 'models/common/text'

import {FC} from 'react'
import classNames from 'classnames';

export interface IPropsText {
  type?: TextTypes,
  weight?: TextWeight,
  size?: TextSize,
  children: string,
  customStyles?: string,
}

export const Text: FC<IPropsText> = ({
  type = TextTypes.p,
  weight,
  size,
  children,
  customStyles,
}) => {
  const classNamesMixed = classNames(
    'text',
    // weight
    {'text--light': weight === TextWeight.LIGHT},
    {'text--regular': weight === TextWeight.REGULAR},
    {'text--medium': weight === TextWeight.MEDIUM},
    {'text--bold': weight === TextWeight.BOLD},
    // sizes
    {'text--xs': size === TextSize.EXTRA_SMALL},
    {'text--sm': size === TextSize.SMALL},
    {'text--md': size === TextSize.MEDIUM},
    {'text--lg': size === TextSize.BIG},
    {'text--xl': size === TextSize.LARGE},
    {'text--xxl': size === TextSize.EXTRA_LARGE},
    // custom
    {customStyles: !!customStyles}
  )

  switch (type) {
    case TextTypes.h1:
      return <h1 className="text text--heading-1">{children}</h1>
    case TextTypes.h2:
      return <h2 className="text text--heading-2">{children}</h2>
    case TextTypes.h3:
      return <h3 className="text text--heading-3">{children}</h3>
    case TextTypes.h4:
      return <h4 className="text text--heading-4">{children}</h4>
    case TextTypes.h5:
      return <h5 className="text text--heading-5">{children}</h5>
    case TextTypes.h6:
      return <h6 className="text text--heading-6">{children}</h6>
    case TextTypes.p:
      return <p className="text text--description">{children}</p>
    case TextTypes.mixed:
      return <p className={classNamesMixed}>{children}</p>
    default:
      return <p className="text text--description">{children}</p>
  }
}
