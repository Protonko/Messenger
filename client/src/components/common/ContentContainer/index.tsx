import type {FC} from 'react'
import {Loader} from 'components/common/Loader'
import {Text} from 'components/common/Text'

interface ContentContainerProps {
  loading: boolean
  errorMessage: string | null
  children: JSX.Element
}

export const ContentContainer: FC<ContentContainerProps> = ({
  loading,
  errorMessage,
  children,
}) => {
  if (errorMessage) {
    return <Text>{errorMessage}</Text>
  }

  if (loading) {
    return <Loader center={true} />
  }

  return children
}
