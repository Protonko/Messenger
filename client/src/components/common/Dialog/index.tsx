import type {FC} from 'react'
import type {IDialog} from 'models/dialog'
import {ReadStatus} from 'models/common/status'
import {TextSize, TextTypes} from 'models/common/text'
import {ReactComponent as DoubleCheck} from 'assets/icons/double-check.svg'
import {ReactComponent as Check} from 'assets/icons/check.svg'
import {COLORS} from 'static/colors'
import {Text} from 'components/common/Text'
import {Counter} from 'components/common/Counter'
import {Avatar} from 'components/common/Avatar'

export const Dialog: FC<IDialog> = ({
  name,
  description,
  avatar,
  edited,
  createdAt,
  updatedAt,
  messages,
  status,
  readStatus,
}) => {
  const counter = messages.toString().length > 2 ? '99+' : messages

  const renderStatus = () => {
    if (!!counter) {
      return <Counter count={counter} status={status}/>
    }

    if (readStatus) {
      const props = {
        width: 15,
        height: 15,
        color: COLORS.dustyGray,
      }
      return readStatus === ReadStatus.READ ? <DoubleCheck {...props} /> : <Check {...props} />
    }

    return null
  }

  return (
    <div className="dialog">
      <Avatar src={avatar} />

      <div className="dialog__data">
        <div className="dialog__text">
          <Text
            type={TextTypes.h4}
            customStyles="dialog__text-title"
            numberOfLines={1}
          >
            {name}
          </Text>
          <Text>{description}</Text>
        </div>

        <div className="dialog__info">
          <Text
            type={TextTypes.mixed}
            size={TextSize.EXTRA_SMALL}
            customStyles="dialog__info-date"
          >
            {new Date(edited ? updatedAt : createdAt).toLocaleDateString()}
          </Text>

          {renderStatus()}
        </div>
      </div>
    </div>
  )
}
