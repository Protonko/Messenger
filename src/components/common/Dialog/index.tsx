// types
import {Status} from 'models/common/status';
import {TextTypes} from 'models/common/text';

import {FC} from 'react'
import {Text} from 'components/common/Text';
import {Counter} from 'components/common/Counter';
import {Avatar} from 'components/common/Avatar';

export const Dialog: FC = () => {
  return (
    <div className="dialog">
        <Avatar src={''} />
        <div className="dialog__text">
          <Text type={TextTypes.h4}>Title</Text>
          <Text>jdifjsifjdfij</Text>
        </div>
        <div className="dialog__info">
          <Text>19:33</Text>
          <Counter count={2} status={Status.MUTED} />
        </div>
    </div>
  )
}
