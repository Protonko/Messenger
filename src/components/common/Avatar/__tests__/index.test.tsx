import {shallow} from 'enzyme'
import image from 'assets/images/test-image.jpg'
import {Avatar, IPropsAvatar} from 'components/common/Avatar'
import {Sizes} from 'models/common/sizes'

const ELEMENT_SELECTORS = {
  avatar: 'avatar',
  avatarSmall: 'avatar--sm',
  avatarLarge: 'avatar--lg',
}

const shallowComponent = (props: IPropsAvatar) => (
  shallow(<Avatar {...props} />)
)

describe('Avatar', () => {
  let props: IPropsAvatar

  beforeEach(() => {
    props = {
      src: image,
      size: Sizes.MEDIUM
    }
  })

  it('Should render Avatar component', () => {
    const component = shallowComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.avatar}`)
    expect(wrapper.length).toBe(1)
  })

  it('Match snapshot', () => {
    const component = shallowComponent(props)

    expect(component).toMatchSnapshot()
  })

  it('Shouldn`t find sm and lg modifiers', () => {
    props.size = undefined

    const component = shallowComponent(props)
    const wrapperSmall = component.find(`.${ELEMENT_SELECTORS.avatarSmall}`)
    const wrapperLarge = component.find(`.${ELEMENT_SELECTORS.avatarLarge}`)
    expect(wrapperSmall.length + wrapperLarge.length).toBe(0)
  })

  it('Should return small classname', () => {
    props.size = Sizes.SMALL

    const component = shallowComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.avatarSmall}`)
    expect(wrapper.length).toBe(1)
  })

  it('Should return large classname', () => {
    props.size = Sizes.LARGE

    const component = shallowComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.avatarLarge}`)
    expect(wrapper.length).toBe(1)
  })
})
