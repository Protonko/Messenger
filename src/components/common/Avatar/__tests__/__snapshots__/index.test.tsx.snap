// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Match snapshot 1`] = `
<div
  className="avatar"
>
  <img
    alt=""
    className="avatar__image"
    src={
      Object {
        "process": [Function],
      }
    }
  />
  <Loader />
</div>
`;
